ReactFlux/scripts/stores/TodoStore.js

```diff
@@ -0,0 +1,125 @@
+import _ from 'lodash';
+import AppDispatcher from '../dispatcher/AppDispatcher';
+import TodoConstants from '../constants/TodoConstants';
+import EventEmitter from 'events';
+
+const ActionTypes = TodoConstants.ActionTypes;
+var CHANGE_EVENT = 'change';
+
+var _todos = {};
+
+function create(text) {
+  var id = (+new Date() + Math.floor(Math.random() * 999999)).toString(36);
+  _todos[id] = {
+    id: id,
+    complete: false,
+    text: text
+  };
+}
+
+function update(id, updates) {
+  _todos[id] = _.assign({}, _todos[id], updates);

```

Comment:
A big problem here (and unfortunately I'm sure we are guilty of this in many places in Reports) is that this store is not immutable - i.e. in this case you are replacing a single value that is part of the `_todos` object, but the reference to the `_todos` object is staying the same. If you compare that to the Redux example, they are recreating the entire state of the "store" (for a given reducer) every time a single item needs to be updated (e.g. by using a `map` on the state array, which creates a new array containing all the old objects).

This has no impact as long as all our React components are based on the `Component` class - but the `Component` class does not give us good performance. Every time something changes in the store, any component that depends on that store has to do a deep comparison against the previous state of the store to figure out whether it needs to re-render itself. And you can imagine stores can get pretty big quite quickly, so the problem gets exponentially worse as more components and stores are introduced! 😅

However, if our stores are immutable (i.e. we recreate the store every time an action changes part of it), we can safely change all of our React components to be based on the `PureComponent` class. The `PureComponent` class is the same as a normal `Component` class, but it includes a `shouldComponentUpdate` method that only does a shallow state comparison to determine whether it should re-render itself - that is to say that it just checks whether the reference to the store has changed, and if it hasn't, then it won't re-render.